src/ORM/Association/BelongsToMany.php

```diff
@@ -797,14 +797,19 @@ protected function _saveLinks(EntityInterface $sourceEntity, $targetEntities, $o
             $sourceKeys = array_combine($foreignKey, $sourceEntity->extract($bindingKey));
             $targetKeys = array_combine($assocForeignKey, $e->extract($targetPrimaryKey));
 
-            if ($sourceKeys !== $joint->extract($foreignKey)) {
-                $joint->set($sourceKeys, ['guard' => false]);
+            $changedKeys = (
+                $sourceKeys !== $joint->extract($foreignKey) ||
+                $targetKeys !== $joint->extract($assocForeignKey)
+            );
+            // Keys were changed, the junction table record _could_ be
```

Comment:
IMO this should be added way more often into the code. Sometimes you really have to dig through the code to understand things.